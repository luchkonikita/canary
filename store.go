package main

import (
	"errors"
	"log"
	"net/http"
	"time"

	"github.com/asdine/storm"
	"github.com/asdine/storm/q"
	"github.com/gorilla/schema"
)

var decoder = schema.NewDecoder()

// DB Entities

type RequestHeader struct {
	Name  string
	Value string
}

// Crawling - a single crawling action, with page results related to it
type Crawling struct {
	ID          int             `schema:"-" storm:"id,increment"`
	CreatedAt   time.Time       `schema:"-"`
	URL         string          `schema:"url" storm:"index"`
	Processed   bool            `schema:"-" storm:"index" `
	Concurrency int             `schema:"concurrency"`
	Headers     []RequestHeader `schema:"headers"`
}

// PageResult - an entity representing a particular requested page
type PageResult struct {
	ID         int `storm:"id,increment"`
	CrawlingID int `storm:"index"`
	URL        string
	Status     int `storm:"index"`
}

// Validate - return an error if crawling is invalid
func (cr *Crawling) Validate() error {
	// TODO: Use regexp to match URL structure
	if len(cr.URL) == 0 {
		return errors.New("Crawling should have a valid URL")
	}
	if cr.Concurrency == 0 {
		return errors.New("Crawling should have a concurrency bigger than 0")
	}
	return nil
}

// DB initialization

// NewDB - initializes a DB and ensures all the buckets are in place
// If resetData is true this will clean up database after initialization (used in tests).
func NewDB(filename string, resetData bool) *storm.DB {
	db, err := storm.Open(filename)
	if err != nil {
		log.Fatal(err)
	}

	if resetData {
		db.Drop(&Crawling{})
		db.Drop(&PageResult{})
	}
	db.Init(&Crawling{})
	db.Init(&PageResult{})

	return db
}

// Actions
// Different functions for mutating data.

// CreateCrawling - creates a new crawling and prepares results for all
// the pages. Results are processed later in a background job.
// If there is a crawling already running for this sitemap, the new one
// will not be created and return will be returned instead.
func CreateCrawling(db *storm.DB, cr *Crawling, r *http.Request) error {
	cr.CreatedAt = time.Now()
	decoder.Decode(cr, r.Form)

	if err := cr.Validate(); err != nil {
		return err
	}

	query := db.Select(q.And(
		q.Eq("URL", cr.URL),
		q.Eq("Processed", false),
	))

	if query.First(&Crawling{}) == nil {
		return errors.New("Current crawling is in progress, cannot create another")
	}

	tx, err := db.Begin(true)
	if err != nil {
		return err
	}

	defer tx.Rollback()

	urls, err := ParseSitemap(cr)
	if err != nil {
		return err
	}

	tx.Save(cr)
	for _, url := range urls {
		tx.Save(&PageResult{
			CrawlingID: cr.ID,
			URL:        url,
			Status:     0,
		})
	}

	return tx.Commit()
}

// DeleteCrawling - deletes a crawling and all related page results
func DeleteCrawling(db *storm.DB, crawlingID int) error {
	var crawling Crawling

	if err := db.One("ID", crawlingID, &crawling); err != nil {
		return err
	}

	var pageResults []PageResult
	db.Find("CrawlingID", crawling.ID, &pageResults)

	tx, _ := db.Begin(true)
	tx.DeleteStruct(&crawling)
	for _, pageResult := range pageResults {
		tx.DeleteStruct(&pageResult)
	}
	return tx.Commit()
}
