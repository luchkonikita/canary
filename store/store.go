package store

import (
	"errors"
	"log"
	"net/http"
	"strconv"
	"time"

	"github.com/asdine/storm"
	"github.com/asdine/storm/q"
)

const queryLimit = 10000

// DB Entities

// Sitemap - an entity representing a particular sitemap
type Sitemap struct {
	ID          int    `storm:"id,increment"`
	Name        string `storm:"unique"`
	URL         string `storm:"unique"`
	Concurrency int
	Username    string
	Password    string
}

// Crawling - a single crawling action, with page results related to it
type Crawling struct {
	ID        int `storm:"id,increment"`
	SitemapID int `storm:"index"`
	CreatedAt time.Time
	Processed bool `storm:"index"`
}

// PageResult - an entity representing a particular requested page
type PageResult struct {
	ID         int `storm:"id,increment"`
	CrawlingID int `storm:"index"`
	URL        string
	Status     int `storm:"index"`
}

// HasAuth - shows if sitemap has basic auth credentials
func (s Sitemap) HasAuth() bool {
	return len(s.Username) > 0 && len(s.Password) > 0
}

// DB initialization

// NewDB - initializes a DB and ensures all the buckets are in place
// If resetData is true this will clean up database after initialization (used in tests).
func NewDB(filename string, resetData bool) *storm.DB {
	db, err := storm.Open(filename)
	if err != nil {
		log.Fatal(err)
	}

	if resetData {
		db.Drop(&Sitemap{})
		db.Drop(&Crawling{})
		db.Drop(&PageResult{})
	}
	db.Init(&Sitemap{})
	db.Init(&Crawling{})
	db.Init(&PageResult{})

	return db
}

// Actions
// Different functions for mutating data.

// CreateSitemap - creates a sitemap
func CreateSitemap(db *storm.DB, sitemap *Sitemap, r *http.Request) error {
	assignSitemapValues(sitemap, r)
	if err := validateSitemap(sitemap); err != nil {
		return err
	}
	return db.Save(sitemap)
}

// UpdateSitemap - updates a sitemap
func UpdateSitemap(db *storm.DB, sitemap *Sitemap, r *http.Request) error {
	assignSitemapValues(sitemap, r)
	if err := validateSitemap(sitemap); err != nil {
		return err
	}
	return db.Update(sitemap)
}

func assignSitemapValues(sitemap *Sitemap, r *http.Request) {
	if concurrency, err := strconv.Atoi(r.FormValue("concurrency")); err == nil {
		sitemap.Concurrency = concurrency
	}
	if _, ok := r.Form["name"]; ok {
		sitemap.Name = r.FormValue("name")
	}
	if _, ok := r.Form["url"]; ok {
		sitemap.URL = r.FormValue("url")
	}
	if _, ok := r.Form["username"]; ok {
		sitemap.Username = r.FormValue("username")
	}
	if _, ok := r.Form["password"]; ok {
		sitemap.Password = r.FormValue("password")
	}
}

func validateSitemap(sitemap *Sitemap) error {
	if len(sitemap.Name) == 0 || len(sitemap.URL) == 0 {
		return errors.New("Sitemap Name and URL cannot be empty")
	}
	return nil
}

// DeleteSitemap - deletes a sitemap found by ID
func DeleteSitemap(db *storm.DB, id int) error {
	var sitemap Sitemap
	if err := db.One("ID", id, &sitemap); err != nil {
		return err
	}

	var crawlings []Crawling
	db.Find("SitemapID", sitemap.ID, &crawlings)

	tx, _ := db.Begin(true)
	tx.DeleteStruct(&sitemap)
	for _, crawling := range crawlings {
		var pageResults []PageResult
		db.Find("CrawlingID", crawling.ID, &pageResults)
		tx.DeleteStruct(&crawling)
		for _, pageResult := range pageResults {
			tx.DeleteStruct(&pageResult)
		}
	}
	return tx.Commit()
}

// CreateCrawling - creates a new crawling and prepares results for all
// the pages. Results are processed later in a background job.
// If there is a crawling already running for this sitemap, the new one
// will not be created and return will be returned instead.
func CreateCrawling(db *storm.DB, sitemap *Sitemap, urls []string) (*Crawling, error) {
	crawling := &Crawling{}
	query := db.Select(q.And(
		q.Eq("SitemapID", sitemap.ID),
		q.Eq("Processed", false),
	))
	err := query.First(&Crawling{})

	if err == nil {
		return crawling, errors.New("Current crawling is in progress, cannot create another")
	}

	tx, err := db.Begin(true)
	if err != nil {
		return crawling, err
	}

	crawling.SitemapID = sitemap.ID
	crawling.CreatedAt = time.Now()
	crawling.Processed = false

	tx.Save(crawling)

	if err != nil {
		return crawling, err
	}
	defer tx.Rollback()

	for _, url := range urls {
		tx.Save(&PageResult{
			CrawlingID: crawling.ID,
			URL:        url,
			Status:     0,
		})
	}

	return crawling, tx.Commit()
}

func DeleteCrawling(db *storm.DB, crawlingID int) error {
	var crawling Crawling

	if err := db.One("ID", crawlingID, &crawling); err != nil {
		return err
	}

	var pageResults []PageResult
	db.Find("CrawlingID", crawling.ID, &pageResults)

	tx, _ := db.Begin(true)
	tx.DeleteStruct(&crawling)
	for _, pageResult := range pageResults {
		tx.DeleteStruct(&pageResult)
	}
	return tx.Commit()
}

// Queries
// Basic query methods without any sophisticated logic

// GetSitemap - retrieves a sitemap by ID
func GetSitemap(db *storm.DB, sitemapID int) (*Sitemap, error) {
	sitemap := &Sitemap{}
	err := db.One("ID", sitemapID, sitemap)
	return sitemap, err
}

// GetSitemaps - retrieves all sitemaps for the store
func GetSitemaps(db *storm.DB) []Sitemap {
	var sitemaps []Sitemap
	db.All(&sitemaps)
	return sitemaps
}

// Filters
// For building advanced queries and using parameters

// CrawlingsFilter a struct used for filtering crawlings based on JSON encoded parameters
type CrawlingsFilter struct {
	Request *http.Request
}

// PageResultsFilter a struct used for filtering page results based on JSON encoded parameters
type PageResultsFilter struct {
	Request *http.Request
}

// Query - applies CrawlingsFilter and returns results
func (f *CrawlingsFilter) Query(db *storm.DB) []Crawling {
	crawlings := []Crawling{}
	filters := []q.Matcher{}

	if sitemapID, err := intValue(f.Request, "sitemap_id"); err == nil {
		filters = append(filters, q.Eq("SitemapID", sitemapID))
	}
	if processed, err := boolValue(f.Request, "processed"); err == nil {
		filters = append(filters, q.Eq("Processed", processed))
	}

	query := db.Select(q.And(filters...))
	query = query.Limit(limitValue(f.Request))
	query = query.Skip(offsetValue(f.Request))

	query.Find(&crawlings)
	return crawlings
}

// Query - applies PageResultsFilter and returns results
func (f *PageResultsFilter) Query(db *storm.DB) []PageResult {
	pageResults := []PageResult{}
	filters := []q.Matcher{}

	if crawlingID, err := intValue(f.Request, "crawling_id"); err == nil {
		filters = append(filters, q.Eq("CrawlingID", crawlingID))
	}
	if status, err := intValue(f.Request, "status"); err == nil {
		filters = append(filters, q.Eq("Status", status))
	}
	if url, err := stringValue(f.Request, "url"); err == nil {
		filters = append(filters, q.Re("URL", url))
	}

	query := db.Select(q.And(filters...))
	query = query.Limit(limitValue(f.Request))
	query = query.Skip(offsetValue(f.Request))

	query.Find(&pageResults)
	return pageResults
}

// Query helper functions
func stringValue(r *http.Request, name string) (string, error) {
	val := r.FormValue(name)
	if val == "" {
		return val, errors.New("is empty")
	}
	return val, nil
}

func intValue(r *http.Request, name string) (int, error) {
	val := r.FormValue(name)
	return strconv.Atoi(val)
}

func boolValue(r *http.Request, name string) (bool, error) {
	val := r.FormValue(name)

	if val == "true" {
		return true, nil
	} else if val == "false" {
		return false, nil
	}
	return false, errors.New("is empty")
}

func offsetValue(r *http.Request) int {
	offset, _ := intValue(r, "offset")
	return offset
}

func limitValue(r *http.Request) int {
	if limit, err := intValue(r, "limit"); err == nil {
		if limit > 0 && limit < queryLimit {
			return limit
		}
		return queryLimit
	}
	return queryLimit
}
