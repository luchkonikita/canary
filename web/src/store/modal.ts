type ShowModalAction = StoreAction<'SHOW_MODAL', { modalType: ModalState['modalType'], modalParams: ModalState['modalParams'] }>
type HideModalAction = StoreAction<'HIDE_MODAL', {}>
type SetModalLoadingAction = StoreAction<'SET_MODAL_LOADING', { loading: boolean }>

type ModalAction = ShowModalAction | HideModalAction | SetModalLoadingAction

export function showModal(type: StoreState['modal']['modalType'], params: any): ShowModalAction {
  return {
    type: 'SHOW_MODAL',
    payload: {
      modalType: type,
      modalParams: params
    }
  }
}

export function hideModal(): HideModalAction {
  return {
    type: 'HIDE_MODAL',
    payload: {}
  }
}

export function setModalLoading(loading: boolean): SetModalLoadingAction {
  return {
    type: 'SET_MODAL_LOADING',
    payload: {loading}
  }
}

export function reducer(state = { show: false, loading: false }, action: ModalAction): ModalState {
  switch (action.type) {
    case 'SHOW_MODAL':
      return {...action.payload, show: true, loading: false}
    case 'HIDE_MODAL':
      return {show: false, loading: false}
    case 'SET_MODAL_LOADING':
      return {...state, loading: action.payload.loading}
    default:
      return state
  }
}
