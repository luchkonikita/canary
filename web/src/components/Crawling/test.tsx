import * as React from 'react'
import { shallow } from 'enzyme'
import { Crawling } from '.'
import { loadCrawling, showModal } from '../../store'

test('Crawling renders', () => {
  const crawling: ICrawling = {
    id: 1,
    url: 'http://example.com/sitemap.xml',
    createdAt: '2018-10-22T21:06:21.765601129+04:00',
    processed: false,
    pageResults: []
  }

  const props = {
    crawlingId: 1,
    loadCrawling,
    showModal,
    crawling
  }

  const rendered = shallow(<Crawling {...props} />)
  expect(rendered).toMatchSnapshot()
})
