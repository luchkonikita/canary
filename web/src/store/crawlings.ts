import { Dispatch } from 'redux'
import { omit } from 'lodash'
import { stringify } from 'qs'
import { toaster } from 'evergreen-ui'

import { setModalLoading, hideModal } from './modal'

function denormalizeCrawlings(crawlings: ICrawling[]): CrawlingsState {
  return crawlings.reduce((memo: any, current: ICrawling) => {
    memo[current.id] = current
    return memo
  }, {})
}

type StoreCrawlingsAction = StoreAction<'STORE_CRAWLINGS', { crawlings: ICrawling[] }>
type StoreCrawlingAction = StoreAction<'STORE_CRAWLING', { crawling: ICrawling }>
type RemoveCrawlingAction = StoreAction<'REMOVE_CRAWLING', { crawlingId: number }>

type CrawlingsAction = StoreCrawlingsAction | StoreCrawlingAction | RemoveCrawlingAction

function storeCrawlings(crawlings: ICrawling[]): StoreCrawlingsAction {
  return { type: 'STORE_CRAWLINGS', payload: { crawlings } }
  }

function storeCrawling(crawling: ICrawling): StoreCrawlingAction {
  return { type: 'STORE_CRAWLING', payload: { crawling } }
  }

function removeCrawling(crawlingId: number): RemoveCrawlingAction {
  return { type: 'REMOVE_CRAWLING', payload: { crawlingId } }
}

export function loadCrawlings() {
  return async function (dispatch: Dispatch) {
    const result = await fetch('http://localhost:4000/crawlings')
    const crawlings = await result.json()
    dispatch(storeCrawlings(crawlings))
  }
}

export function loadCrawling(id: number) {
  return async function (dispatch: Dispatch) {
    const result = await fetch(`http://localhost:4000/crawlings/${id}`)
    const crawling = await result.json()
    return dispatch(storeCrawling(crawling))
  }
}

export function deleteCrawling(id: number) {
  return async function (dispatch: Dispatch) {
    dispatch(setModalLoading(true))
    const result = await fetch(`http://localhost:4000/crawlings/${id}`, { method: 'DELETE' })

    if (result.status === 200) {
      dispatch(removeCrawling(id))
      dispatch(hideModal())
    } else {
      toaster.danger('Something went wrong')
      dispatch(setModalLoading(false))
    }
  }
}

export function createCrawling(data: {url: string, concurrency: number, headers: {name: string, value: string}[]}) {
  return async function (dispatch: Dispatch) {
    dispatch(setModalLoading(true))

    const params: {[index: string]: any} = {
      url: data.url,
      concurrency: data.concurrency,
    }

    data.headers.forEach((header, index) => {
      params[`headers.${index}.name`] = header.name
      params[`headers.${index}.value`] = header.value
    })

    const body = new URLSearchParams(stringify(params))

    const result = await fetch('http://localhost:4000/crawlings', { method: 'POST', body })

    if (result.status === 201) {
      const crawling = await result.json()

      dispatch(storeCrawling(crawling))
      dispatch(hideModal())
    } else {
      const message = await result.text()
      toaster.danger(`Failed to start: ${message}`)
      dispatch(setModalLoading(false))
    }
  }
}

export function reducer(state: CrawlingsState = [], action: CrawlingsAction): CrawlingsState {
  switch (action.type) {
    case 'STORE_CRAWLINGS':
      return denormalizeCrawlings(action.payload.crawlings)
    case 'STORE_CRAWLING':
      return { ...state, ...denormalizeCrawlings([action.payload.crawling]) }
    case 'REMOVE_CRAWLING':
      return omit(state, action.payload.crawlingId)
    default:
      return state
  }
}
