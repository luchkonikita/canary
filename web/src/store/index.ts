import { createStore, applyMiddleware, combineReducers } from 'redux'
import thunkMiddleware from 'redux-thunk'

import { reducer as crawlingsReducer } from './crawlings'
import { reducer as modalReducer } from './modal'

export { loadCrawlings, loadCrawling, createCrawling, deleteCrawling } from './crawlings'
export { showModal, hideModal } from './modal'

export default createStore(
  combineReducers<StoreState>({
    crawlings: crawlingsReducer,
    modal: modalReducer
  }),
  applyMiddleware(thunkMiddleware)
)
