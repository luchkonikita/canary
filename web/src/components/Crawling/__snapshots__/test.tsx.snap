// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crawling renders 1`] = `
<Pane>
  <Pane
    alignItems="center"
    display="flex"
    justifyContent="space-between"
    padding={16}
  >
    <Pane
      alignItems="center"
      display="flex"
    >
      <Pane
        alignItems="center"
        display="flex"
        height={20}
        justifyContent="center"
        width={20}
      >
        <StatusIcon
          type="none"
        />
      </Pane>
      <Pane
        display="flex"
        flexDirection="column"
        marginLeft={16}
      >
        <Text
          color="default"
          cursor="pointer"
          fontFamily="ui"
          fontWeight={500}
          onClick={[Function]}
          size={500}
          textStyles={
            Object {
              "100": Object {
                "fontSize": "10px",
                "fontWeight": 400,
                "lineHeight": "16px",
              },
              "200": Object {
                "fontSize": "11px",
                "fontWeight": 400,
                "lineHeight": "16px",
              },
              "300": Object {
                "fontSize": "12px",
                "fontWeight": 400,
                "letterSpacing": "0",
                "lineHeight": "16px",
              },
              "400": Object {
                "fontSize": "14px",
                "fontWeight": 400,
                "letterSpacing": "-0.05px",
                "lineHeight": "22px",
              },
              "500": Object {
                "fontSize": "16px",
                "fontWeight": 400,
                "letterSpacing": "-0.05px",
                "lineHeight": "22px",
              },
              "600": Object {
                "fontSize": "18px",
                "fontWeight": 400,
                "letterSpacing": "-0.05px",
                "lineHeight": "24px",
              },
              "700": Object {
                "fontSize": "20px",
                "fontWeight": 400,
                "letterSpacing": "-0.05px",
                "lineHeight": "24px",
              },
              "800": Object {
                "fontSize": "24px",
                "fontWeight": 400,
                "letterSpacing": "-0.2px",
                "lineHeight": "28px",
              },
              "900": Object {
                "fontSize": "29px",
                "fontWeight": 400,
                "letterSpacing": "-0.2px",
                "lineHeight": "32px",
              },
            }
          }
          textUppercaseStyles={
            Object {
              "100": Object {
                "fontSize": "10px",
                "fontWeight": 400,
                "letterSpacing": "1px",
                "lineHeight": "16px",
                "textTransform": "uppercase",
              },
              "200": Object {
                "fontSize": "11px",
                "fontWeight": 400,
                "letterSpacing": "1px",
                "lineHeight": "16px",
                "textTransform": "uppercase",
              },
            }
          }
          title="Click to copy the URL"
        >
          http://example.com/sitemap.xml
        </Text>
        <Text
          color="default"
          fontFamily="ui"
          marginTop={2}
          size={300}
          textStyles={
            Object {
              "100": Object {
                "fontSize": "10px",
                "fontWeight": 400,
                "lineHeight": "16px",
              },
              "200": Object {
                "fontSize": "11px",
                "fontWeight": 400,
                "lineHeight": "16px",
              },
              "300": Object {
                "fontSize": "12px",
                "fontWeight": 400,
                "letterSpacing": "0",
                "lineHeight": "16px",
              },
              "400": Object {
                "fontSize": "14px",
                "fontWeight": 400,
                "letterSpacing": "-0.05px",
                "lineHeight": "22px",
              },
              "500": Object {
                "fontSize": "16px",
                "fontWeight": 400,
                "letterSpacing": "-0.05px",
                "lineHeight": "22px",
              },
              "600": Object {
                "fontSize": "18px",
                "fontWeight": 400,
                "letterSpacing": "-0.05px",
                "lineHeight": "24px",
              },
              "700": Object {
                "fontSize": "20px",
                "fontWeight": 400,
                "letterSpacing": "-0.05px",
                "lineHeight": "24px",
              },
              "800": Object {
                "fontSize": "24px",
                "fontWeight": 400,
                "letterSpacing": "-0.2px",
                "lineHeight": "28px",
              },
              "900": Object {
                "fontSize": "29px",
                "fontWeight": 400,
                "letterSpacing": "-0.2px",
                "lineHeight": "32px",
              },
            }
          }
          textUppercaseStyles={
            Object {
              "100": Object {
                "fontSize": "10px",
                "fontWeight": 400,
                "letterSpacing": "1px",
                "lineHeight": "16px",
                "textTransform": "uppercase",
              },
              "200": Object {
                "fontSize": "11px",
                "fontWeight": 400,
                "letterSpacing": "1px",
                "lineHeight": "16px",
                "textTransform": "uppercase",
              },
            }
          }
        >
          21:06 / October 22
        </Text>
      </Pane>
    </Pane>
    <Pane
      display="flex"
    >
      <Button
        appearance="default"
        height={32}
        iconAfterAim="none"
        iconBeforeAim="none"
        isActive={false}
        onClick={[Function]}
        paddingBottom={0}
        paddingTop={0}
      >
        View Details
      </Button>
      <Button
        appearance="default"
        height={32}
        iconAfterAim="none"
        iconBeforeAim="none"
        isActive={false}
        marginLeft={16}
        onClick={[Function]}
        paddingBottom={0}
        paddingTop={0}
      >
        Delete
      </Button>
    </Pane>
  </Pane>
</Pane>
`;
